components/properties/src/provider.rs

```diff
@@ -311,11 +317,152 @@ impl<'data, T: TrieValue> PropertyCodePointMapV1<'data, T> {
+/// including in SemVer minor releases. While the serde representation of data structs is guaranteed
+/// to be stable, their Rust representation might not be. Use with caution.
+/// </div>
+#[derive(PartialEq, Eq)] // VarULE wants these to be byte equality
+#[derive(VarULE)]
+#[repr(transparent)]
+pub struct NormalizedPropertyNameStr([u8]);
+
+#[cfg(feature = "datagen")]
+impl serde::Serialize for NormalizedPropertyNameStr {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use serde::ser::Error;
+        if serializer.is_human_readable() {
+            let s = str::from_utf8(&self.0)
+                .map_err(|_| S::Error::custom("Attempted to datagen invalid string property"))?;
+            serializer.serialize_str(s)
+        } else {
+            serializer.serialize_bytes(&self.0)
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for Box<NormalizedPropertyNameStr> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use alloc::borrow::Cow;
+        let s; // lifetime lengthening
+        let b;
+        // Can be improved with https://github.com/unicode-org/icu4x/issues/2310
+        // the allocations here are fine, in normal ICU4X code they'll only get hit
+        // during human-readable deserialization
+        let bytes = if deserializer.is_human_readable() {
+            s = <Cow<str>>::deserialize(deserializer)?;
+            s.as_bytes()
+        } else {
+            b = <Cow<[u8]>>::deserialize(deserializer)?;
+            &b
+        };
+        Ok(NormalizedPropertyNameStr::boxed_from_bytes(bytes))
+    }
+}
+
+impl<'a> ZeroMapKV<'a> for NormalizedPropertyNameStr {
+    type Container = VarZeroVec<'a, NormalizedPropertyNameStr>;
+    type Slice = VarZeroSlice<NormalizedPropertyNameStr>;
+    type GetType = NormalizedPropertyNameStr;
+    type OwnedType = Box<NormalizedPropertyNameStr>;
+}
+
+impl PartialOrd for NormalizedPropertyNameStr {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Normalize a character based on the "loose matching" described in PropertyValueAliases.txt,
+/// returning None for skippable characters
+///
+/// ICU has [code for this][1] (and [during property lookup][2]) which we emulate.
+/// In particular, ICU only does normalization within ASCII, which makes sense since character names
+/// seem to be only ASCII.
+///
+/// [1]: https://github.com/unicode-org/icu/blob/288c4c7555915ce7b1fb675d94ddd495058fc039/icu4c/source/common/propname.cpp#L35
+/// [2]: https://github.com/unicode-org/icu/blob/288c4c7555915ce7b1fb675d94ddd495058fc039/icu4c/source/common/propname.cpp#L226-L230
+fn normalize_char(ch: u8) -> Option<u8> {
+    match ch {
+        b'a'..=b'z' => Some(ch),
+        b'A'..=b'Z' => Some(ch | 0x20), // Uppercase by setting last bit
```

Comment:
Do you mean "lowercase by setting first bit?". Anyway, use `to_ascii_uppercase` instead of bit twiddling.